README
astahttpd v0.1 RC1 README file
-------------------------------

I wrote astahttpd back in early 2008 as a fun project and stop developing it
at the same year. At that time PHP did not have built-in web server. I don't
have plan neither to make new features nor do bug fixing on astahttpd.

What is astahttpd?
------------------

astahttpd is free and open source web server written using 100% pure PHP.
Currently astahttpd v0.1-RC1 support these following features:

   - GET, POST, and HEAD methods
   - Alias directories
   - Modular Architecture
   - CGI Processing (mod_cgi & mod_cgi_header)
   - Virtual Host (mod_vhost)
   - URL Rewrite (mod_rewrite)
   - Basic Authentication  (mod_auth_basic)
   - Digest Authentication (mod_auth_digest)
   - Log written to file (mod_log *)
   - Bandwidth Limiting (mod_bandwidth)
   - Live server status (mod_status)
   - Content encoding using gzip or deflate (mod_encoding)
   - Directory Browsing (mod_dir_browser *)
   - Win32 Support *
   - Simple caching (mod_cache) *
   - Autofix URI (mod_autofix_uri) *
   - Static content (mod_static *)

Note: * new in this release

You need at least PHP 5.2.3 to run astahttpd since it implements Object
Oriented features that only available on PHP5.

astahttpd SHOULD NOT be used in production environment due many performance and
security issue that has not been fixed yet.

The Idea
--------

1. The idea behind the making of astahttpd is simple, I just need simple
web server for astasms (an sms server written also in pure PHP). Since
astasms front-end is web based, so it needs a web server. I don't want
to use Apache or other mainstream web server player, because i think
it's too heavy if used only for supporting front-end.

   After tried few simple httpd server written by somebody else, but most of them
just make me more frustating. Many of them is really hard to install and
configure and the worst, i even cannot compile them :( .

   So I started to make my own web server using my favorite language PHP. The first
thnik I need to decide is whether to use PHP4 or PHP5. I think OOP in PHP5
is very good and make software maintanance easier. So I decided to use PHP5.

2. Just for fun :) .

Current Version
---------------

Current version that shipped with this distribution is astahttpd v0.1-RC1

Latest Release
--------------

You can obtain astahttpd latest release on project homepage at
https://github.com/astasoft/astahttpd

Installation
------------

Quick start

```
$ git clone git@github.com:astasoft/astahttpd
$ cd astahttpd
$ cp conf/aws.conf.php.sample conf/aws.conf.php
$ php bin/aws
```

Open your web browser and point to http://localhost:8000

For details see file called INSTALL for Linux and INSTALL_WIN32 for
Windows users.

License
-------

astahttpd is licensed under GNU GPL v3, for more information see file
called LICENSE.txt

Author
-------
Rio Astamal (me@rioastamal.net)

NEWS
01-Apr-2008, astahttpd 0.1 RC1
-------------------------------
- Win32 Support
- Some part now moved as module, such as:
  - mod_static
  - mod_log
  - mod_dir_browser
- New modules:
  - mod_autofix_uri
  - mod_cache ( HTTP 304 Not Modified )
- astahttpd does not need pcntl extension anymore
- Dropped support for IP Based Virtual Host
- many improvements
- Fixed bug #1929009 (Internal gzip compression not work)
- Fixed bug #1929215 (Could not handle same header name in HTTP Header)

07-Feb-2008, astahttpd 0.1 beta3
---------------------------------
- Digest Authentication
- Modular Architecture
- Currenly available modules: mod_auth_basic, mod_auth_digest, mod_bandwidth,
  mod_cgi_header, mod_cgi, mod_encoding, mod_rewrite, mod_status, mod_vhost
- Fixed bug #1887560 (CGI header did not handled properly)
- Fixed bug #1885235 (Error not properly handled on external gzip)
- Fixed bug #1885216 (Missing compression level on external gzip)

01-Feb-2008, astahttpd 0.1 beta2
--------------------------------
- Fixed bug #1882558, Always return Error 500 when listing a directory
- Added Manual/API
- Added Simple URL Rewriting
- Added deflate format for Content-Encoding
  (PHP must be compiled using --with-zlib)
- Added Basic Authentication
- License changed from GPLv2 to GPLv3
- New comment style to suit PHPDocumentor style guide
- Log written to file (access_log & rewrite_log)

26-Jan-2008, astahttpd 0.1 beta1
--------------------------------
- Fixed bug #1879557, (Not currently handlle \r\n\r\n characters on client request)
- Added both IP and name based Virtual Host Support
- Added Content-Encoding support (gzip)

23-Jan-2008, astahttpd 0.1 alpha
--------------------------------
- First public release
- GET, HEAD, POST methods
- CGI processing
- Alias Directories support


modules/mod_cgi.php
<?php
/**
 * @filesource
 */
 
/**
 * Module for providing CGI script processing
 * - Created on Thu, 07 Feb 2008 21:31:13 GMT+7
 * - Updated on Mon, 31 Mar 2008 22:16 GMT+7 
 *
 * @package       astahttpd
 * @subpackage    modules
 * @copyright     Copyright (c) 2008, Rio Astamal
 * @author        Rio Astamal <me@rioastamal.net>
 * @version       0.1 (for astahttpd >= v0.1-beta3)
 * @link          http://astahttpd.sourceforge.net/
 * @license       http://opensource.org/licenses/gpl-license.php GNU GPLv3
 * @since         0.1-beta3
 *
 */
 
final class mod_cgi extends Module {

   /**
    * Class constructor
    */
   public function __construct() {
      parent::__construct();
      $this->modName = 'mod_cgi';
      $this->modVer = '0.1';
      $this->modAuthor = 'Rio Astamal';
      $this->modCycle = 'second_cycle';
      $this->modType = 'Standard';
      $this->modPriority = 4;
      $this->modDesc = 'Module for providing CGI script processing';
   }
   
   /**
    * Main method to activate module
    *
    * @return void
    */    
   public function activate() {
      if (!is_dir($this->V_Target)) {
         $htparser =& $this->O_HParser;
         $htbuilder =& $this->O_HMaker;
         $server =& $this->O_HttpServer;
         $cgi =& $this->O_Cgi;
         $content =& $this->V_Content;
         $cgiHeader =& $this->V_CgiHeader;
         $staticContent =& $this->V_StaticContent;
         $cgiEnv =& $this->V_CgiEnv;
         
         $target = $this->V_Target;
         
         if (strstr($server->getCgiExt(), $htparser->getFileExtension()) || 
            !$htparser->getFileExtension()) {
            
            $staticContent = false; // change the flag
            
            $reqtype = $htparser->getRequestType();
            $executor = $server->setExecuteHandlerFor($htparser->getFileExtension());
            
            if (!is_executable($target) && !$htparser->getFileExtension()) {
               throw new HttpException (
                  "You do not have permission to execute this file.\n",
                  'Error 403 - Access Forbidden',
                  403
               );
            }
                  
            // set all environment variables for this script
            HttpServer::liveDebug("INSIDE MOD_CGI: Value of PATH_INFO IS ".
               $htparser->getPathInfo()."\n");
            
            $cgiEnv['DOCUMENT_ROOT'] = realpath($server->getBaseDir());
            $cgiEnv['GATEWAY_INTERFACE'] = 'CGI/1.1';
            $cgiEnv['PATH'] = $_ENV['PATH'];
            $cgiEnv['PATH_TRANSLATED'] = $htparser->getPathTranslated();
            $cgiEnv['PATH_INFO'] = $htparser->getPathInfo();
            $cgiEnv['QUERY_STRING'] = $htparser->getQueryString();         
            $cgiEnv['REMOTE_ADDR'] = $server->getPeerName();
            $cgiEnv['REQUEST_URI'] = $htparser->getRequestUri();
            $cgiEnv['REQUEST_METHOD'] = $reqtype;
            $cgiEnv['SERVER_SOFTWARE'] = $server->getServerFullName();
            $cgiEnv['SERVER_NAME'] = $server->getHostName();
            $cgiEnv['SERVER_ADDR'] = '127.0.0.1'; // $server->getHost();
            $cgiEnv['SERVER_PORT'] = $server->getListenPort();
            $cgiEnv['SERVER_SIGNATURE'] = $server->getFooter();
            $cgiEnv['SERVER_PROTOCOL'] = 'HTTP/1.1';
            $cgiEnv['SERVER_ADMIN'] = $server->getServerAdmin();         
            $cgiEnv['SCRIPT_FILENAME'] = $target;
            $cgiEnv['SCRIPT_NAME'] = $htparser->getScriptName();
            $cgiEnv['REDIRECT_STATUS'] = '200';
            
            if (IS_WINDOWS) {
               // it's sucks, made me headeach :)
               foreach ($_ENV as $key=>$val) {
                 $cgiEnv[strtoupper($key)] = $val;
               }
            }
                           
            foreach ($htparser->getAllHeader() as $header=>$val) {
               $header = 'HTTP_'.strtoupper(str_replace('-', '_', $header));
               $cgiEnv[$header] = $val;
            }
            
            $postdata = null;
            $proc = null;
            $pipo = null;
            $resp = '';
            $normalProc = true;
            
            // black lists file
            $black_lists = array('wp-cron.php');   // Wordpress wp-cron.php
                     
            if ($reqtype == 'POST') {
               $cgiEnv['CONTENT_LENGTH'] = $htparser->getHeader('Content-Length');
               // something like application/x-www-form-urlencoded
               $cgiEnv['CONTENT_TYPE'] = $htparser->getHeader('Content-Type');
               $postdata = $htparser->getPostData();
               $desc = array(
                  0 => array('pipe', 'r'), // STDIN
                  1 => array('pipe', 'w'), // STDOUT
                  2 => array('pipe', 'w')  // STDERR
               );              
            } else {
               $desc = array( 
                  1 => array('pipe', 'w'), // STDIN
                  2 => array('pipe', 'w')  // STDERR
               );
            }
                     
            $_current = basename($target);
            
            // HttpServer::liveDebug(print_r($cgiEnv, true));
            
            // I DON'T KNOW WHY IF WE PROVIDE ACTUAL QUERY STRING TO wp-cron.php
            // IT BECOME HANG
            /**
             * TODO: FIX hang wp-cron.php
             */         
            if (in_array($_current, $black_lists) != false) {
               HttpServer::liveDebug("__ACTUAL QUERY_STRING: {$cgiEnv['QUERY_STRING']}\n");
               $cgiEnv['QUERY_STRING'] = 'what-happen-to-query-string--i-dont-know-why';
            }
            
            if (IS_WINDOWS) {
               $bypass = array('bypass_shell' => true);
               $proc = proc_open($executor.' '.$target, $desc, $pipo, null, $cgiEnv, $bypass);
            } else {
               $proc = proc_open($executor.' '.$target, $desc, $pipo, null, $cgiEnv);
            }
            var_dump($proc);
            if (!is_resource($proc)) {
               throw new HttpException("proc_open() failure.", "Error 500 - Internal Server Error", 500);
            }
            
            if ($reqtype == 'POST') {
               fwrite($pipo[0], $postdata);
               fclose($pipo[0]);
            }
            
            $buf = '';
                  
            //do {
               // stream_set_blocking($pipo[1], 0);
               
               // var_dump($pipo[1]);
               // $meta = stream_get_meta_data($pipo[1]);
               
               // print_r($meta);            
               $resp = stream_get_contents($pipo[1]);
               
               // stream_set_timeout($pipo[1], 1, 500000);
               // print "__TRYING FOR: ".(++$zero)." times\n";
               
               //$buf = fread($pipo[1], 4096);
               // print "____CURRENT BUF: $buf\n";
               //$resp .= $buf;
               
            //} while ($buf);
            
            fclose($pipo[1]);
            $err = proc_close($proc);
            
            // var_dump($err, $fc);
            unset($proc, $cgiEnv, $pipo, $desc);
            
            $cgiHeader = $htbuilder->getCgiHeader($resp, $content);
            // print "CGI HEADER: $cgiHeader\nCONTENT: $content\n";
         }
      }
   }
}
